src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+ *   cross-entropy. If negative or zero, 500 epochs are used if the graph is the graph is small
+ *   (less than 50k edges), 50 epochs are used for larger graphs.
+ * \param sampling_prob The fraction of vertices moved at each iteration of the stochastic gradient descent (epoch). At fixed number of epochs, a higher fraction makes the algorithm slower. Vice versa, a too low number will converge very slowly, possibly too slowly. If negative of zero, a value of 0.3 is assumed.
+ *
+ * \return Error code.
+ *
+ */
+igraph_error_t igraph_layout_umap(const igraph_t *graph, const igraph_vector_t *distances,
+    igraph_matrix_t *layout, const igraph_real_t min_dist, const igraph_integer_t epochs, const igraph_real_t sampling_prob);
+
+
+/* rho is just the size of the distance from each vertex and its closest neighbor */
+/* sigma is the the decay from each vertex, depends on its rho and the rest of its neighbor
+ * distances */
+
+/* Find sigma for this vertex by binary search */
+static igraph_error_t igraph_i_umap_find_sigma(const igraph_t *graph,
+        const igraph_vector_t *distances, igraph_integer_t i, const igraph_vector_int_t *eids,
+        igraph_real_t rho, igraph_real_t *sigma_p, igraph_real_t target) {
+
+    igraph_real_t sigma = 1;
+    igraph_real_t sum;
+    igraph_real_t tol = 0.01;
+    igraph_integer_t maxiter = 100;
+    igraph_integer_t no_of_neis = igraph_vector_int_size(eids);
+    igraph_integer_t eid;
+    igraph_real_t step = sigma;
+    int seen_max = 0;
+
+    /* Binary search */
+    for(int iter = 0; iter < maxiter; iter++) {
+        sum = 0;
+        for(int j = 0; j < no_of_neis; j++) {
+            eid = VECTOR(*eids)[j];
+            sum += exp(-(VECTOR(*distances)[eid] - rho) / sigma);
+        }
+
+#ifdef UMAP_DEBUG
+        printf("SIGMA function (i = %" IGRAPH_PRId ", no_of_neis = %" IGRAPH_PRId ")- sum: %f, "
+               "target: %f, rho: %f, sigma: %f\n", i, no_of_neis, sum, target, rho, sigma);
+#endif
+
+        /* TODO: this seems fine, but is probably a little off for some corner cases */
+        if (sum < target) {
+            /* going back up after having seen an upper bound */
+            if (seen_max == 1) {
+                step /= 2;
+            /* we need to go up but have not seen an upper bound yet
+             * first iteration we want to increase by sigma, else we must come from
+             * below, so we are sitting at 2 * step, we want to move to 4 * step */
+            } else if (iter > 0) {
+                step *= 2;
+            }
+            sigma += step;
+        /* overshooting, we have definitely seen the max */
+        } else {
+            seen_max = 1;
+            step /= 2;
+            sigma -= step;
+        }
+
+        /* Check for convergence */
+        if (fabs(sum - target) < tol) {
+            break;
+        }
+    }
+
+    *sigma_p = sigma;
+
+    return (IGRAPH_SUCCESS);
```

Comment:
For consistent formatting:

```suggestion
    return IGRAPH_SUCCESS;
```

`return` is not a function.

Reply:
That was probably my fault, I'm usually working on a project where we have pretty weird formatting requirements, and `return (something)` is one of them.